libs/wingc/src/visit.rs

```diff
@@ -0,0 +1,272 @@
+use crate::ast::{
+	ArgList, Class, Constructor, Expr, ExprKind, FunctionDefinition, InterpolatedStringPart, Literal, Reference, Scope,
+	Stmt, StmtKind,
+};
+
+/// Visitor pattern inspired by implementation from https://docs.rs/syn/latest/syn/visit/index.html
+///
+/// If you are implementing a visitor, you can delegate back to the
+/// default implementation to visit the children of a node by calling
+/// visit::visit_<node_type>(self, node);
+///
+/// For example:
+///
+/// ```ignore
+/// impl<'ast> Visit<'ast> for ExprVisitor {
+///   fn visit_item_fn(&mut self, exp: &'ast Expr) {
+///     println!("Expr with span={}", exp.span);
+///
+///     // Delegate to the default impl to visit any nested expressions.
+///     visit::visit_expr(self, exp);
+///   }
+/// }
+/// ```
+pub trait Visit<'ast> {
+	fn visit_scope(&mut self, e: &'ast Scope) {
```

Comment:
Would be cool if we could find some way to pass some [`FnOnce`](https://doc.rust-lang.org/std/ops/trait.FnOnce.html) here so we can just provide a closure of "what to do" in this node that returns a `bool` telling the default visitor implementation if we want to continue to traverse deeper into this node or return. This way we don't need to copy the default implementation for each visit function we create.

Reply:
I think this is a good option to have, but sometimes closures are more annoying to deal with (due to borrowing rules) so it shouldn't be the only way. I originally implemented something like this (visitor with FnMut) in my hover PR and ended up getting rid of it.